### PR TITLE
Add `stdlib/Michelson`

### LINKED ISSUE #468: Add Michelson primitives file
`stdlib/Michelson.ju` - should expose all the primitive types from [here](https://hackage.haskell.org/package/morley-1.5.0/docs/Michelson-Untyped-Type.html) and functions from [here](https://hackage.haskell.org/package/morley-1.5.0/docs/Michelson-Untyped-Instr.html) (though not all instructions are functions) using the `%prim` syntax. Then we can import this file in contracts which compile to the Michelson backend.

It would be nice if Juvix can automatically expose the stdlib (whenever `juvix` is invoked), but this can happen later.

Per @mariari best to import qualified for now.

### PR DESCRIPTION
**Pull request by cyberglot**
Closes #468.

### CODE REVIEW

**Comment by cyberglot** [comment]
Some types are not clear to me but I will go back to them after I'm done with the whole list of functions.

**Comment by cwgoes** [comment]
We need to add all the arithmetic operations, which should be dependently-typed (eventually we'll use typeclasses).

**Comment by cyberglot** [comment]
`add : x : Type -> x -> x -> x` like this or?

**Comment by cwgoes** [comment]
> `add : x : Type -> x -> x -> x` like this or?

Yes, essentially, later on it will be restricted to typeclasses.

File: stdlib/Michelson.ju [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+    type hash-key	 
+    type timestamp	 
+    type address
+
+    -- instructions
+    -- not clear what the types are from documentation (more research needed)
+    -- `words` are typed as `bytes`
+    -- otherwise `a` 
+    sig ext : a -- ???
+    sig dropn : int -> bytes -- return type???
```
Comment (cwgoes) [inline stdlib/Michelson.ju:35]: We won't have these operations (anything which directly manipulates the stack)

Comment (cwgoes) [reply stdlib/Michelson.ju:35]: (there's no stack in the Juvix semantics, we just compile the lambda calculus to it for Michelson)

File: stdlib/Michelson.ju [hunk 2]
```diff
@@ -0,0 +1,140 @@
...
+    let sub = %Michelson.sub
+
+    sig div : (x: type) -> x -> x -> x
+    let div = %Michelson.div
+
+    sig abs : (x: type) -> x -> x -> x
+    let abs = %Michelson.abs
+
+    -- boolean functions
+    sig and : (x: type) -> x -> x -> x
```
Comment (mariari) [inline stdlib/Michelson.ju:47]: And will be reserved

File: stdlib/Michelson.ju [hunk 3]
```diff
@@ -0,0 +1,140 @@
...
+    let chain-id = %Michelson.chain-id
+    let int = %Michelson.int
+    let nat = %Michelson.nat
+    let string = %Michelson.string
+    let bytes = %Michelson.bytes
+    let mutez = %Michelson.mutez
+    let bool = %Michelson.bool
+    let hash-key = %Michelson.hash-key
+    let timestamp = %Michelson.timestamp
+    let address = %Michelson.address
```
Comment (mariari) [inline stdlib/Michelson.ju:28]: If we have a bunch of these in a row like this, can we align the `=` ?

File: stdlib/Michelson.ju [hunk 4]
```diff
@@ -0,0 +1,140 @@
+mod Michelson =
```
Comment (mariari) [inline stdlib/Michelson.ju:1]: We don't need this mod at the top...

Though we will once we have a top level thing I supsect it'll look like

```haskelll
module Michelson where
...
```

Though for now, I'll just use the filename and capitalize the first

File: stdlib/Michelson.ju [hunk 5]
```diff
@@ -0,0 +1,143 @@
...
+  sig amount : (x : let) -> x -> ()
+  let amount = %Michelson.amount
+
+  sig balance : (x : let) -> address -> x
+  let balance = %Michelson.balance
+
+  sig sha256 : (x: type) -> x -> hash-key
+  let sha256 = %Michelson.sha256
+
+  sig sha512 :
```
Comment (cwgoes) [inline stdlib/Michelson.ju:73]: This and subsequent hash functions are `bytes -> bytes`

File: stdlib/Michelson.ju [hunk 6]
```diff
@@ -0,0 +1,143 @@
...
+  sig source : address
+  let source = %Michelson.source
+
+  sig sender : address
+  let sender = %Michelson.sender
+
+  sig implicit-account : address
+  let implicit-account = %Michelson.implicit-account
+
+  sig create-contract : () -> signature
```
Comment (cwgoes) [inline stdlib/Michelson.ju:88]: This & transfer-tokens create an operation

File: stdlib/Michelson.ju [hunk 7]
```diff
@@ -0,0 +1,143 @@
...
+  sig empty-set : set a
+  let empty-set = %Michelson.empty-set
+
+  sig empty-map : map a
+  let empty-map = %Michelson.empty-map
+
+  sig empty-big-map : big-map a
+  let empty-big-map = %Michelson.empty-big-map
+
+  sig map : (x: type) -> lambda a b -> x a -> x b
```
Comment (cwgoes) [inline stdlib/Michelson.ju:133]: The compiler could do the `lambda` wrapping automatically but maybe explicit is better (like you have here)

Comment (mariari) [reply stdlib/Michelson.ju:133]: ahh I missed thing, `(a -> b)` works for lambda

File: stdlib/Michelson.ju [hunk 8]
```diff
@@ -0,0 +1,143 @@
...
+  let balance = %Michelson.balance
+
+  sig sha256 : (x: type) -> x -> hash-key
+  let sha256 = %Michelson.sha256
+
+  sig sha512 :
+  let sha512 = %Michelson.sha512
+
+  sig blake2b :
+  let blake2b = %Michelson.
```
Comment (mariari) [inline stdlib/Michelson.ju:77]: This neeeds to be finished

File: stdlib/Michelson.ju (hunk 6 above)
Comment (mariari) [inline stdlib/Michelson.ju:88]: seems like we bind unit, probably best to use that for now

File: stdlib/Michelson.ju [hunk 9]
```diff
@@ -0,0 +1,143 @@
...
+  sig transfer-tokens : (x: let) -> address -> x -> hash-key
+  let transfer-token = %Michelson.transfer-tokens
+
+  sig set-delegate : address -> ()
+  let set-delegate = %Michelson.set-delegate
+
+  sig none : either () a
+  let none = %Michelson.none
+
+  sig some : b -> either a b
```
Comment (mariari) [inline stdlib/Michelson.ju:100]: Same here

Comment (mariari) [reply stdlib/Michelson.ju:100]: probably best to use b for the second argument of either (doesn't matter, just a nit!)

File: stdlib/Michelson.ju [hunk 10]
```diff
@@ -0,0 +1,143 @@
...
+  let empty-big-map = %Michelson.empty-big-map
+
+  sig map : (x: type) -> lambda a b -> x a -> x b
+  let map = %Michelson.map
+
+  sig update : (x: type) -> key -> a -> x a -> ()
+  let update = %Michelson.update
+
+  sig iter : (x: type) -> lambda a b -> lambda b ()
+  let iter = %Michelson.iter
```
Comment (mariari) [inline stdlib/Michelson.ju:140]: same here

**Comment by mariari** [comment]
One last change, we can use normal lambdas for michelson, so you can convert lambda a b to (a -> b)

**Comment by cyberglot** [comment]
> One last change, we can use normal lambdas for michelson, so you can convert lambda a b to (a -> b)

will do.

**Review by cwgoes — APPROVED** [review]
ACK, we'll need to write lots of tests

**Pull request merged**